File: videoframes/src/main/java/com/atmaca/videokareleri/MainActivity.java
package com.atmaca.videokareleri;

import android.Manifest;
import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends Activity {
    private static final int REQ_TREE = 1907;
    private static final int REQ_NOTIFY = 1908;
    private Uri selectedTree;
    private TextView folderText;
    private TextView statusText;
    private ProgressBar progress;
    private Button startButton;

    private final BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override public void onReceive(Context context, Intent intent) {
            if (!FrameExtractService.ACTION_PROGRESS.equals(intent.getAction())) return;
            String message = intent.getStringExtra(FrameExtractService.EXTRA_MESSAGE);
            int done = intent.getIntExtra(FrameExtractService.EXTRA_DONE, 0);
            int total = intent.getIntExtra(FrameExtractService.EXTRA_TOTAL, 0);
            statusText.setText(message == null ? "Çalışıyor" : message);
            progress.setIndeterminate(total <= 0);
            if (total > 0) {
                progress.setMax(total);
                progress.setProgress(Math.min(done, total));
            }
            boolean finished = intent.getBooleanExtra(FrameExtractService.EXTRA_FINISHED, false);
            startButton.setEnabled(finished || selectedTree != null);
        }
    };

    @Override protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        buildUi();
        String saved = getPreferences(MODE_PRIVATE).getString("tree", null);
        if (saved != null) {
            selectedTree = Uri.parse(saved);
            folderText.setText(selectedTree.toString());
            startButton.setEnabled(true);
        }
        if (Build.VERSION.SDK_INT >= 33 && checkSelfPermission(Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.POST_NOTIFICATIONS}, REQ_NOTIFY);
        }
    }

    private void buildUi() {
        int pad = (int) (20 * getResources().getDisplayMetrics().density);
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(pad, pad, pad, pad);
        root.setGravity(Gravity.CENTER_HORIZONTAL);

        TextView title = new TextView(this);
        title.setText("Video Kareleri");
        title.setTextSize(28f);
        title.setGravity(Gravity.CENTER);
        root.addView(title, new LinearLayout.LayoutParams(-1, -2));

        TextView desc = new TextView(this);
        desc.setText("Seçilen klasördeki videolardan her saniye 1 JPEG çıkarır. Başarılı video silinir; hata veren video HATA klasörüne taşınır.");
        desc.setTextSize(16f);
        desc.setPadding(0, pad, 0, pad);
        root.addView(desc, new LinearLayout.LayoutParams(-1, -2));

        Button select = new Button(this);
        select.setText("Klasör Seç");
        select.setOnClickListener(v -> chooseTree());
        root.addView(select, new LinearLayout.LayoutParams(-1, -2));

        folderText = new TextView(this);
        folderText.setText("Klasör seçilmedi");
        folderText.setPadding(0, pad / 2, 0, pad);
        root.addView(folderText, new LinearLayout.LayoutParams(-1, -2));

        startButton = new Button(this);
        startButton.setText("Başlat");
        startButton.setEnabled(false);
        startButton.setOnClickListener(v -> startWork());
        root.addView(startButton, new LinearLayout.LayoutParams(-1, -2));

        progress = new ProgressBar(this, null, android.R.attr.progressBarStyleHorizontal);
        progress.setIndeterminate(false);
        LinearLayout.LayoutParams pp = new LinearLayout.LayoutParams(-1, -2);
        pp.setMargins(0, pad, 0, pad / 2);
        root.addView(progress, pp);

        statusText = new TextView(this);
        statusText.setText("Hazır");
        statusText.setTextSize(16f);
        root.addView(statusText, new LinearLayout.LayoutParams(-1, -2));

        setContentView(root);
    }

    private void chooseTree() {
        Intent i = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
        i.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION | Intent.FLAG_GRANT_PREFIX_URI_PERMISSION);
        startActivityForResult(i, REQ_TREE);
    }

    @Override protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode != REQ_TREE || resultCode != RESULT_OK || data == null || data.getData() == null) return;
        selectedTree = data.getData();
        int flags = data.getFlags() & (Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
        try { getContentResolver().takePersistableUriPermission(selectedTree, flags); } catch (Exception ignored) {}
        getPreferences(MODE_PRIVATE).edit().putString("tree", selectedTree.toString()).apply();
        folderText.setText(selectedTree.toString());
        startButton.setEnabled(true);
        statusText.setText("Klasör hazır");
    }

    private void startWork() {
        if (selectedTree == null) return;
        Intent i = new Intent(this, FrameExtractService.class);
        i.setAction(FrameExtractService.ACTION_START);
        i.putExtra(FrameExtractService.EXTRA_TREE_URI, selectedTree.toString());
        if (Build.VERSION.SDK_INT >= 26) startForegroundService(i); else startService(i);
        startButton.setEnabled(false);
        statusText.setText("Videolar hazırlanıyor…");
        progress.setIndeterminate(true);
    }

    @Override protected void onStart() {
        super.onStart();
        IntentFilter filter = new IntentFilter(FrameExtractService.ACTION_PROGRESS);
        if (Build.VERSION.SDK_INT >= 33) registerReceiver(receiver, filter, Context.RECEIVER_NOT_EXPORTED);
        else registerReceiver(receiver, filter);
    }

    @Override protected void onStop() {
        super.onStop();
        try { unregisterReceiver(receiver); } catch (Exception ignored) {}
    }
}

File: vk5/app/src/main/kotlin/com/videokareleri/v5/MainActivity.kt
package com.videokareleri.v5

import android.Manifest
import android.app.*
import android.content.*
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.widget.*
import kotlinx.coroutines.*
import java.util.LinkedHashSet

class MainActivity : Activity() {
    private val selected = LinkedHashSet<String>()
    private lateinit var folders: TextView
    private lateinit var status: TextView
    private lateinit var progress: ProgressBar
    private lateinit var choose: Button
    private lateinit var start: Button
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(c: Context?, i: Intent?) {
            i ?: return
            val v=i.getIntExtra("videos",0); val t=i.getIntExtra("total",0)
            val s=i.getIntExtra("saved",0); val k=i.getIntExtra("skipped",0); val e=i.getIntExtra("errors",0)
            val cur=i.getStringExtra("current").orEmpty(); val fin=i.getBooleanExtra("finished",false)
            progress.isIndeterminate=t<=0
            if(t>0){ progress.max=t; progress.progress=v }
            status.text="$cur\nVideo: $v/$t   Kaydedilen: $s   Zaten var: $k   Hata: $e"
            if(fin){ start.isEnabled=true; choose.isEnabled=true }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        buildUi(); loadSelection(); ensurePermission()
    }

    private fun buildUi() {
        val navy=Color.rgb(0,35,102); val yellow=Color.rgb(255,220,0)
        val root=LinearLayout(this).apply { orientation=LinearLayout.VERTICAL; setPadding(dp(20),dp(24),dp(20),dp(20)); setBackgroundColor(navy) }
        root.addView(TextView(this).apply { text="VİDEO KARELERİ V5"; textSize=26f; setTextColor(yellow); setTypeface(null,1) }, LinearLayout.LayoutParams(-1,-2))
        root.addView(TextView(this).apply { text="Native Kotlin • Hafif mod • Her saniyeden 1 JPEG"; setTextColor(Color.WHITE); textSize=14f; setPadding(0,dp(8),0,dp(18)) })
        choose=button("VİDEO KLASÖRLERİNİ SEÇ",yellow,navy).also { it.setOnClickListener { chooseFolders() } }
        root.addView(choose,LinearLayout.LayoutParams(-1,dp(54)))
        folders=TextView(this).apply { setTextColor(Color.WHITE); textSize=15f; setPadding(0,dp(14),0,dp(14)) }
        root.addView(folders)
        start=button("KARELERİ ÇIKAR",yellow,navy).also { it.setOnClickListener { startJob() } }
        root.addView(start,LinearLayout.LayoutParams(-1,dp(54)))
        val stop=button("DURDUR",Color.WHITE,navy).apply { setOnClickListener { startService(Intent(this@MainActivity,FrameExtractService::class.java).setAction(FrameExtractService.ACTION_CANCEL)) } }
        root.addView(stop,LinearLayout.LayoutParams(-1,dp(50)).apply { topMargin=dp(10) })
        progress=ProgressBar(this,null,android.R.attr.progressBarStyleHorizontal)
        root.addView(progress,LinearLayout.LayoutParams(-1,dp(12)).apply { topMargin=dp(20) })
        status=TextView(this).apply { text="Hazır"; setTextColor(Color.WHITE); textSize=15f; setPadding(0,dp(14),0,0) }
        root.addView(status)
        setContentView(root)
    }

    private fun button(text:String,bg:Int,fg:Int)=Button(this).apply { this.text=text; setTextColor(fg); textSize=16f; setTypeface(null,1); setBackgroundColor(bg) }
    private fun requiredPermission()=if(Build.VERSION.SDK_INT>=33) Manifest.permission.READ_MEDIA_VIDEO else Manifest.permission.READ_EXTERNAL_STORAGE
    private fun hasPermission()=checkSelfPermission(requiredPermission())==PackageManager.PERMISSION_GRANTED
    private fun ensurePermission(){ if(!hasPermission()) requestPermissions(arrayOf(requiredPermission()),REQ_MEDIA) }

    private fun chooseFolders() {
        if(!hasPermission()){ ensurePermission(); return }
        choose.isEnabled=false; status.text="Video klasörleri sayılıyor..."
        scope.launch {
            val map=withContext(Dispatchers.IO){ MediaStoreVideoRepository(contentResolver).loadFolderCounts() }
            choose.isEnabled=true; showFolderDialog(map)
        }
    }

    private fun showFolderDialog(map: LinkedHashMap<String,Int>) {
        if(map.isEmpty()){ status.text="Erişilebilir video bulunamadı."; return }
        val paths=map.keys.toTypedArray(); val labels=Array<CharSequence>(paths.size){ i -> "${paths[i]}  (${map[paths[i]]})" }; val checked=BooleanArray(paths.size){ selected.contains(paths[it]) }
        AlertDialog.Builder(this).setTitle("Video klasörlerini seç")
            .setMultiChoiceItems(labels,checked){_,which,isChecked-> if(isChecked) selected.add(paths[which]) else selected.remove(paths[which]) }
            .setPositiveButton("TAMAM"){_,_->saveSelection();refreshFolders()}.setNegativeButton("İPTAL",null).show()
    }

    private fun startJob() {
        if(selected.isEmpty()){ Toast.makeText(this,"Önce en az bir klasör seç",Toast.LENGTH_SHORT).show(); return }
        val i=Intent(this,FrameExtractService::class.java).setAction(FrameExtractService.ACTION_START).putExtra(FrameExtractService.EXTRA_PATHS,selected.toTypedArray())
        if(Build.VERSION.SDK_INT>=26) startForegroundService(i) else startService(i)
        start.isEnabled=false; choose.isEnabled=false; status.text="İşlem başlatıldı..."
    }

    private fun saveSelection()=getSharedPreferences("vk5",MODE_PRIVATE).edit().putStringSet("folders",selected).apply()
    private fun loadSelection(){ selected.clear(); selected.addAll(getSharedPreferences("vk5",MODE_PRIVATE).getStringSet("folders",emptySet()).orEmpty()); refreshFolders() }
    private fun refreshFolders(){ folders.text=if(selected.isEmpty()) "Seçili klasör: yok" else "Seçili klasör: ${selected.size}" }

    override fun onStart(){ super.onStart(); val f=IntentFilter(FrameExtractService.ACTION_PROGRESS); if(Build.VERSION.SDK_INT>=33) registerReceiver(receiver,f,RECEIVER_NOT_EXPORTED) else registerReceiver(receiver,f) }
    override fun onStop(){ runCatching{unregisterReceiver(receiver)}; super.onStop() }
    override fun onDestroy(){ scope.cancel(); super.onDestroy() }
    private fun dp(v:Int)=Math.round(v*resources.displayMetrics.density)
    companion object { const val REQ_MEDIA=41 }
}
